read-fonts/src/tables/colr.rs

```diff
@@ -101,4 +157,686 @@ impl<'a> Colr<'a> {
+            }
+        }
+    }
+}
+
+struct Colrv1ClosureContext<'a> {
+    glyph_set: IntSet<GlyphId>,
+    layer_indices: &'a mut IntSet<u32>,
+    palette_indices: &'a mut IntSet<u16>,
+    variation_indices: &'a mut IntSet<u32>,
+    colr: &'a Colr<'a>,
+    nesting_level_left: u8,
+    visited_paints: IntSet<u32>,
+    colr_head: usize,
+}
+
+impl<'a> Colrv1ClosureContext<'a> {
+    pub fn new(
+        layer_indices: &'a mut IntSet<u32>,
+        palette_indices: &'a mut IntSet<u16>,
+        variation_indices: &'a mut IntSet<u32>,
+        colr: &'a Colr,
+    ) -> Self {
+        let colr_head = colr.offset_data().as_bytes().as_ptr() as usize;
+        Self {
+            glyph_set: IntSet::empty(),
+            layer_indices,
+            palette_indices,
+            variation_indices,
+            colr,
+            nesting_level_left: 64,
+            visited_paints: IntSet::empty(),
+            colr_head,
+        }
+    }
+
+    pub fn dispatch(&mut self, paint: &Paint) {
+        if self.nesting_level_left == 0 {
+            return;
+        }
+
+        if self.paint_visited(paint) {
+            return;
+        }
+        self.nesting_level_left -= 1;
+        paint.v1_closure(self);
+        self.nesting_level_left += 1;
+    }
+
+    fn paint_visited(&mut self, paint: &Paint) -> bool {
+        let offset = (paint.offset_data().as_bytes().as_ptr() as usize - self.colr_head) as u32;
+        if self.visited_paints.contains(offset) {
+            return true;
+        }
+
+        self.visited_paints.insert(offset);
+        false
+    }
+
+    pub fn add_layer_indices(&mut self, first_layer_index: u32, last_layer_index: u32) {
+        self.layer_indices
+            .insert_range(first_layer_index..=last_layer_index);
+    }
+
+    pub fn add_palette_index(&mut self, palette_index: u16) {
+        self.palette_indices.insert(palette_index);
+    }
+
+    pub fn add_variation_indices(&mut self, var_index_base: u32, num_vars: u8) {
+        if num_vars == 0 || var_index_base == NO_VARIATION_INDEX {
+            return;
+        }
+
+        let last_var_index = var_index_base + num_vars as u32 - 1;
+        self.variation_indices
+            .insert_range(var_index_base..=last_var_index);
+    }
+
+    pub fn add_glyph_id(&mut self, gid: GlyphId16) {
+        self.glyph_set.insert(GlyphId::from(gid));
+    }
+}
+
+impl ColorStop {
+    fn v1_closure(&self, c: &mut Colrv1ClosureContext) {
+        c.add_palette_index(self.palette_index());
+    }
+}
+
+impl VarColorStop {
+    fn v1_closure(&self, c: &mut Colrv1ClosureContext) {
+        c.add_palette_index(self.palette_index());
+        c.add_variation_indices(self.var_index_base(), 2);
+    }
+}
+
+impl<'a> ColorLine<'a> {
+    fn v1_closure(&self, c: &mut Colrv1ClosureContext) {
+        for colorstop in self.color_stops() {
+            colorstop.v1_closure(c);
+        }
+    }
+}
+
+impl<'a> VarColorLine<'a> {
+    fn v1_closure(&self, c: &mut Colrv1ClosureContext) {
```

Comment:
this is starting to look an awful lot like a trait...

Reply:
yes, I assume this is fine cause we don't need to use trait bound in functions